### PR TITLE
Add a mechanism for automatically signing out users

### LINKED ISSUE #67: Add an optional auto-signout mechanism
Add a mechanism to automatically sign out users at the end of the event time + post event active time. They should only be credited with the original event end time and not the post event active time. 

Bonus points for adding a configurable mechanism that deletes the active stamps at the end of the post event active time without giving students credit.

And a default empty configuration that does nothing.

### PR DESCRIPTION
**Pull request by bot190**
Adds a scheduled job that will sign out (or discard) active entries when an event's grace period ends. The behavior can be adjusted in the config file by setting AUTO_SIGNOUT_BEHAVIOR.  "Credit": Gives the user credit for the original event end time "Discard": Deletes the active entry without giving the user credit

Any other value will result in nothing happening to expired active entries. 

This mechanism is implemented using Flask-APScheduler which provides access to the app context from within a scheduled job running in a separate thread.

### CODE REVIEW

File: signinapp/__init__.py [hunk 1]
```diff
@@ -202,6 +222,15 @@ def init_default_db():
             end=now + offset,
             type_=db.session.scalar(select(EventType).filter_by(name="Build")),
         )
+
+        expired_event = Event(
```
Comment (msoucy) [inline signinapp/__init__.py:222]: Do you need to add this and notTraining to line 234?

File: signinapp/jobs.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                if current_app.config["AUTO_SIGNOUT_BEHAVIOR"] == "Credit":
+                    # Expire the active session
+                    create_stamp(active)
+                elif current_app.config["AUTO_SIGNOUT_BEHAVIOR"] == "Discard":
+                    # Delete active entry without crediting the user
+                    db.session.delete(active)
+                    db.session.commit()
+
+
+def create_stamp(active: Active):
```
Comment (msoucy) [inline signinapp/jobs.py]: This should be moved into the model, and used for the event page

File: signinapp/jobs.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+@scheduler.task("interval", id="UserSignOutJob", seconds=30)
+def EventEndJob():
+    """
+    This monitor checks all of the entries in the active table
+        For each entry check if the adjusted end time has passed
+    """
+    with scheduler.app.app_context():
+        time = datetime.now(tz=ZoneInfo(current_app.config["TIME_ZONE"]))
+        active_users: List[Active] = db.session.scalars(select(Active))
```
Comment (msoucy) [inline signinapp/jobs.py]: These aren't users, they're entries

File: signinapp/jobs.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+        for active in active_users:
+            if active.event.adjusted_end < time:
+                if current_app.config["AUTO_SIGNOUT_BEHAVIOR"] == "Credit":
+                    # Expire the active session
+                    create_stamp(active)
+                elif current_app.config["AUTO_SIGNOUT_BEHAVIOR"] == "Discard":
+                    # Delete active entry without crediting the user
+                    db.session.delete(active)
+                    db.session.commit()
```
Comment (msoucy) [inline signinapp/jobs.py]: Do we need to commit in the loop?

File: signinapp/jobs.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
+from datetime import datetime
+from typing import List
+from zoneinfo import ZoneInfo
+
+from flask_apscheduler import APScheduler
+from flask import current_app
+from sqlalchemy.future import select
+
+from signinapp.model import Active, Stamps, db
```
Comment (msoucy) [inline signinapp/jobs.py]: Please use `from .model import x`, relative imports

File: signinapp/__init__.py [hunk 6]
```diff
@@ -35,11 +48,15 @@ class Config(object):
     SECRET_KEY = "1234"
     PRE_EVENT_ACTIVE_TIME = 30
     POST_EVENT_ACTIVE_TIME = 120
+    AUTO_SIGNOUT_BEHAVIOR = "None"  # Valid Options (Credit, Discard, None)
 
 
 class DebugConfig(Config):
     TITLE = "Chop Shop Sign In (debug)"
     DB_NAME = ":memory:"
+    PRE_EVENT_ACTIVE_TIME = 30
```
Comment (msoucy) [inline signinapp/__init__.py]: These should be unnecessary, since debug config inherits from config

File: signinapp/__init__.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
 import flask_excel as excel
 import yaml
 from flask import Flask, render_template
+from flask_apscheduler import APScheduler
```
Comment (msoucy) [inline signinapp/__init__.py]: Is this import necessary?

File: requirements.txt [hunk 8]
```diff
@@ -9,6 +9,7 @@ dnspython==2.2.1
 email-validator==1.3.0
 Flask==2.2.2
 Flask-Admin==1.6.0
+Flask-APScheduler
```
Comment (msoucy) [inline requirements.txt]: Are there version constraints?

We probably should trim the list to only the direct dependencies and be more general about versions...

**Review by msoucy — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 8 above)
Comment (bot190) [reply requirements.txt]: Did a little digging and didn't find any obvious documentation indicating what versions of flask are supported, etc. I'll pin this at the current version for now, but I agree, we should probably loosen up these requirements at some point.

**Review by msoucy — APPROVED** [review]

File: requirements.txt [hunk 9]
```diff
@@ -9,6 +9,7 @@ dnspython==2.2.1
 email-validator==1.3.0
 Flask==2.2.2
 Flask-Admin==1.6.0
+Flask-APScheduler == 1.12.4
```
Comment (msoucy) [inline requirements.txt]: ❌

**Pull request merged**